### PR TITLE
feat(chart): improve secrets handling for Token and webhook services.

### PR DESCRIPTION
**Pull request by aledegano**
This PR refactors how the two secrets used by the Token Repository service and Webhook service are handled.

Particularly the Helm chart can now address these 3 different use cases:

1. Upgrading an existing Renkulab deployment: the old secrets will be read and ported in the new secret with the correct encoding. Admins are still asked to manually port the secrets in the values file too, for redundancy.
2. Generating random secrets when not explicitly defined in the values file (e.g. a CI or test deployment), while still preserving those secrets if chart is being re-deployed/upgraded
3. Accepting new secrets from the values file in a new Renkulab deployment

It also comes with a few fail safe:

1. If secrets are specified in the values file and the old secrets exist make sure they match (e.g. double checks that the admins will have correctly followed the instructions).
2. If the values file still contain the secret in the old position fail and ask to remove them.

/deploy extra-values=graph.tokenRepository.tokenEncryption=null,graph.webhookService.hookToken=null,global.renku.cli_version=2.7.0

### CODE REVIEW

**Review by jachro — APPROVED** [review]
👍🏻

File: helm-chart/renku/templates/graph/db-encryption-secret.yaml [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+    {{- end -}}
+  {{- end -}}
+
+  {{- $secretName := include "renku.graph.webhookService.fullname" . -}}
+  {{- $oldWebhookSecret := (lookup "v1" "Secret" .Release.Namespace $secretName) -}}
+  {{- if $oldWebhookSecret -}}
+    {{- $newWebhookSecret := index $oldWebhookSecret.data "webhookService-hookToken-secret" | b64dec | b64dec | trim -}}
+    {{/* If a secret has been specified in the values make sure that corresponds to the base64-decoded secret already present in the cluster, or fail */}}
+    {{- if and .Values.graph.webhookService.aesEncryptionKey (ne $newWebhookSecret .Values.graph.webhookService.aesEncryptionKey) -}}
+      {{- fail (printf "The webhook AES encryption key provided in `.graph.webhookService.aesEncryptionKey` does not match with the base64 decoded key present in %s." $secretName) -}}
```
Comment (olevski) [inline helm-chart/renku/templates/graph/db-encryption-secret.yaml:31]: Why not have the same message here as what you have above? In the message above you print the different secret values.

File: helm-chart/values.yaml.changelog.md [hunk 2]
```diff
@@ -5,6 +5,36 @@ For changes that require manual steps other than changing values, please check o
...
+
+To cleanup the old secrets, please remove the following fields:
+
+* DELETE `graph.tokenRepository.tokenEncryption.secret`
+* DELETE `graph.webhookService.hookToken.secret`
+
+Then extract the content of the old secrets for the Token Repository service using `kubectl`:
+
+```shell
+kubectl -n renku get secrets renku-token-repository -o jsonpath="{.data['tokenRepository-tokenEncryption-secret']}" | base64 -d | base64 -d
```
Comment (olevski) [inline helm-chart/values.yaml.changelog.md:23]: The secrets are base64 encoded twice? This is really confusing.

**Comment by olevski** [comment]
Also @aledegano have you tested this when you go from the current stable renku version to this? All the secrets and everything gets migrated properly?

File: helm-chart/renku/templates/graph/db-encryption-secret.yaml [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+{{- $graphEncKeys := lookup "v1" "Secret" .Release.Namespace $secretName -}}
+{{- if $graphEncKeys -}}
+  {{- $tokenRepository = index $graphEncKeys.data "tokenRepository-aes-encryption-key" -}}
+  {{- $webhookAesEncryptionKey = index $graphEncKeys.data "webhook-aes-encryption-key" -}}
+{{/* If it doesn't exist check if it's in the previous form or in the values */}}
+{{- else -}}
+  {{- $secretName := include "renku.graph.tokenRepository.fullname" . -}}
+  {{- $oldTokenSecret := lookup "v1" "Secret" .Release.Namespace $secretName -}}
+  {{- if $oldTokenSecret -}}
+    {{- $newTokenSecret :=  index $oldTokenSecret.data "tokenRepository-tokenEncryption-secret" | b64dec | b64dec | trim -}}
```
[the remark this answers is not in the document]
Comment (jachro) [reply helm-chart/renku/templates/graph/db-encryption-secret.yaml:16]: Yes, that was the bug Alessandro found the other day. The old secrets were base64 encoded twice. This piece is to ensure the old secrets reach KG services in the shape as before.

File: helm-chart/values.yaml.changelog.md [hunk 4]
```diff
@@ -5,6 +5,36 @@ For changes that require manual steps other than changing values, please check o
 Please follow this convention when adding a new row
 * `<type: NEW|EDIT|DELETE> - *<resource name>*: <details>`
 
+## Upgrading to Renku 0.44.0
```
[the remark this answers is not in the document]
Comment (aledegano) [reply helm-chart/values.yaml.changelog.md]: Typo. Fixed, thanks.

File: helm-chart/renku/templates/graph/webhook-service-deployment.yaml [hunk 5]
```diff
@@ -36,8 +36,8 @@ spec:
             - name: HOOK_TOKEN_SECRET
               valueFrom:
                 secretKeyRef:
-                  name: {{ template "renku.graph.webhookService.fullname" . }}
-                  key: webhookService-hookToken-secret
+                  name: {{ template "renku.graph.dbEncryptionSecret.name" . }}
+                  key: webhook-aes-encryption-key
```
[the remark this answers is not in the document]
Comment (aledegano) [reply helm-chart/renku/templates/graph/webhook-service-deployment.yaml:40]: AFAIK they are both encryption keys fed to AES used to encrypt stuff in the DB.

File: helm-chart/values.yaml.changelog.md (hunk 2 above)
Comment (aledegano) [reply helm-chart/values.yaml.changelog.md:25]: Alas they are (you can verify this yourself by issuing the above command in any cluster). This is one of the point of this PR: to fix this tech debt that we carried around for a while...

File: helm-chart/renku/templates/graph/db-encryption-secret.yaml (hunk 1 above)
Comment (aledegano) [reply helm-chart/renku/templates/graph/db-encryption-secret.yaml]: Typo. Fixed, thanks.

File: helm-chart/renku/templates/graph/db-encryption-secret.yaml [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+       {{- fail (printf "The token repository AES encryption key provided in `graph.tokenRepository.aesEncryptionKey` does not match with the base64 decoded key present in %s. %s != %s" $secretName $newTokenSecret (.Values.graph.tokenRepository.aesEncryptionKey | b64enc)) -}}
+    {{- else -}}
+      {{- $tokenRepository = $newTokenSecret | b64enc | quote -}}
+    {{- end -}}
+  {{- end -}}
+
+  {{- $secretName := include "renku.graph.webhookService.fullname" . -}}
+  {{- $oldWebhookSecret := (lookup "v1" "Secret" .Release.Namespace $secretName) -}}
+  {{- if $oldWebhookSecret -}}
+    {{- $newWebhookSecret := index $oldWebhookSecret.data "webhookService-hookToken-secret" | b64dec | b64dec | trim -}}
```
[the remark this answers is not in the document]
Comment (aledegano) [reply helm-chart/renku/templates/graph/db-encryption-secret.yaml:28]: As I mentioned below it is on purpose.

File: helm-chart/values.yaml.changelog.md (hunk 2 above)
Comment (jachro) [reply helm-chart/values.yaml.changelog.md:25]: I know, that was one of the reasons we decided to do the change.

File: helm-chart/renku/templates/graph/db-encryption-secret.yaml (hunk 3 above)
[the remark this answers is not in the document]
Comment (aledegano) [reply helm-chart/renku/templates/graph/db-encryption-secret.yaml:16]: Yes, as explained below.

File: helm-chart/renku/templates/graph/token-repository-secret.yaml [hunk 7]
```diff
@@ -1,12 +0,0 @@
...
-metadata:
-  name: {{ template "renku.graph.tokenRepository.fullname" . }}
-  labels:
-    app: {{ template "renku.graph.tokenRepository.name" . }}
-    chart: {{ template "renku.chart" . }}
-    release: {{ .Release.Name }}
-    heritage: {{ .Release.Service }}
-type: Opaque
-data:
-  tokenRepository-tokenEncryption-secret: {{ required "Fill in .Values.graph.tokenRepository.tokenEncryption.secret with `openssl rand -hex 8|base64`" .Values.graph.tokenRepository.tokenEncryption.secret | b64enc | quote }}
```
Comment (aledegano) [inline helm-chart/renku/templates/graph/token-repository-secret.yaml:12]: @olevski FYI this is why these secrets are currently double-encoded: we ask the admin to generate the secret with: `openssl rand -hex 8|base64` but then the template base64 encodes that **again**.

File: helm-chart/renku/templates/graph/db-encryption-secret.yaml [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+  {{- $webhookAesEncryptionKey = index $graphEncKeys.data "webhook-aes-encryption-key" -}}
+{{/* If it doesn't exist check if it's in the previous form or in the values */}}
+{{- else -}}
+  {{- $secretName := include "renku.graph.tokenRepository.fullname" . -}}
+  {{- $oldTokenSecret := lookup "v1" "Secret" .Release.Namespace $secretName -}}
+  {{- if $oldTokenSecret -}}
+    {{- $newTokenSecret :=  index $oldTokenSecret.data "tokenRepository-tokenEncryption-secret" | b64dec | b64dec | trim -}}
+    {{/* If a secret has been specified in the values make sure that corresponds to the base64-decoded secret already present in the cluster, or fail */}}
+    {{- if and .Values.graph.tokenRepository.aesEncryptionKey (ne $newTokenSecret .Values.graph.tokenRepository.aesEncryptionKey) -}}
+       {{- fail (printf "The token repository AES encryption key provided in `graph.tokenRepository.aesEncryptionKey` does not match with the base64 decoded key present in %s. %s != %s" $secretName $newTokenSecret (.Values.graph.tokenRepository.aesEncryptionKey | b64enc)) -}}
```
[the remark this answers is not in the document]
Comment (aledegano) [reply helm-chart/renku/templates/graph/db-encryption-secret.yaml]: You're right, I've fixed it now: they both show the un-encoded values.

**Comment by aledegano** [comment]
> Also @aledegano have you tested this when you go from the current stable renku version to this? All the secrets and everything gets migrated properly?

yes, I have tested this multiple times (one last time just now), where I deploy `0.42.0`, defining `graph.tokenRepository.tokenEncryption.secret` and `graph.webhookService.hookToken.secret` and then upgrading to this version, trying the various failure paths as well as the happy path.

In all cases it seems to me that the chart behaves as I expect, migrating the secrets when everything is correct.

FYI the cleanup of the old secrets comes for free as Helm will take care of removing them as the templates are also gone.

**Comment by rokroskar** [comment]
If the migration of the secrets breaks for some reason, what is the fallout? Can the two services be recovered?

**Comment by jachro** [comment]
> If the migration of the secrets breaks for some reason, what is the fallout? Can the two services be recovered?

In fact, we don't migrate the secrets. The old secrets are kept as they are and fed into KG services in a single base64 encoded form. The new deployments, though, will get simple ASCII values. KG services are tweaked so they can work with both.

**Comment by rokroskar** [comment]
OK thanks for clarifying Kuba!

**Review by rokroskar — APPROVED** [review]

File: helm-chart/renku/templates/graph/token-repository-secret.yaml (hunk 7 above)
Comment (olevski) [reply helm-chart/renku/templates/graph/token-repository-secret.yaml:12]: Wow that is wild. I never noticed that.

**Review by olevski — APPROVED** [review]

**Review by olevski — APPROVED** [review]

**Review by rokroskar — APPROVED** [review]

**Pull request merged**